Adding Nexia thermostat support (Trane / American Standard)

ryannazaretian opened the pull request:
## Description:

This adds unofficial Nexia support for Trane and American Standard thermostats.

Community members have been testing this for a while now:
https://community.home-assistant.io/t/nexia-and-trane-xl1050-xl850-thermostat/48673

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#9775

## Example entry for `configuration.yaml` (if applicable):
```yaml
nexia:
  username: "YOUR_USERNAME"
  password: "YOUR_PASSWORD"
  id: YOUR_HOUSE_ID
  scan_interval: (optional) SECONDS_BETWEEN_UPDATES
   
climate:
  - platform: nexia
    
binary_sensor:
  - platform: nexia
    
sensor:
  - platform: nexia
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [ ] There is no commented out code in this PR.
     There is some code that was enabled but commented out due to bugs in the Nexia API. 
  - [X] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)



If the code communicates with devices, web services, or third-party tools:
  - [X] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [ ] Untested files have been added to `.coveragerc`.
    No untested files, does not apply. 

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.
    Interacts with devices, does not apply. 

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

ryannazaretian committed:
Ring: Added Ring Flood Light light component.

Source: https://community.home-assistant.io/t/ring-floodlight-cam-turn-floodlight-on-and-off/75761

Full credit to robpfis07.

ryannazaretian committed:
First release for Nexia

Need to add these to get it working:

nexia:
  username: <your email address>
  password: <your password>
  id: <house id>
 # The house id can be found from the mynexia url:
 # https://www.mynexia.com/houses/<your id>/climate

climate:
  - platform: nexia

binary_sensor:
  - platform: nexia

sensor:
  - platform: nexia

ryannazaretian committed:
Fixed device naming, added attribution.

ryannazaretian committed:
Removed device classes for the binary sensors.

ryannazaretian committed:
Code cleanup and many bug fixes.

ryannazaretian committed:
Fixed bug with set_zone_heat_cool_temp

ryannazaretian committed:
Fixed a bug with emergency heat's state not getting updated.

ryannazaretian committed:
Added scan_interval parameter to rate limit updates. Interval is in seconds.
Changed default update rate / scan interval of the NexiaThermostat class to 120 (from 300)
Added support for celsius and the 0.5 degree stepping.
Hopefully fixed status on climate component.
Fixed bug with getting the system status.
Added system status constants... there may be more though.
Added get_requested_comrpessor_speed().
Changed get_compressor_speed to get_current_compressor_speed.
Removed set_preset restrictions to use preset_modes.
Added new sensor components.

ryannazaretian committed:
Added ability to disable auto-update (causing HASS warnings)
Fixed setpoints set when on COOL or HEAT.
Added get_air_cleaner_mode() function to NexiaThermostat

ryannazaretian committed:
Added support for multiple thermostat systems.
Fixed an issue with update() being called several times due to several components using the same device.

ryannazaretian committed:
Added thermostat and zone IDs into the device attributes.

ryannazaretian committed:
PEP8 and inspection fixes. Fails a lot of pylint tests.

ryannazaretian committed:
Added thermostat_id to has_emergency_heat() call.

ryannazaretian committed:
Added thermostat_id to has_*() calls.

ryannazaretian committed:
Added thermostat_id to several has_*() calls.

ryannazaretian committed:
Fixed naming.

docBliny committed:
Fix multiple unit support by adding `thermostat_id` to calls.

* Also fix header comment about supported unit type.

docBliny committed:
Fix setting temperature with multiple units.

ryannazaretian committed:
Merge remote-tracking branch 'origin/dev' into dev

# Conflicts:
#	homeassistant/components/nexia/binary_sensor.py
#	homeassistant/components/nexia/sensor.py

ryannazaretian committed:
Merge pull request #1 from docBliny/dev

Fix multiple unit support by adding `thermostat_id` to calls.

ryannazaretian committed:
Pylint cleanup

ryannazaretian committed:
Merge remote-tracking branch 'origin/dev' into dev

# Conflicts:
#	homeassistant/components/nexia/nexia_thermostat.py
#	homeassistant/components/nexia/sensor.py

ryannazaretian committed:
Pylint cleanup

ryannazaretian committed:
Fixed setting temperature while zone was off. (removed checks)

ryannazaretian committed:
Updated manifest.

ryannazaretian committed:
Added ability to set the air cleaner mode.

ryannazaretian committed:
Removed Ring Light for Nexia PR

ryannazaretian committed:
Added services.yaml to describe the air cleaner mode.

ryannazaretian committed:
Updated requirements_all.txt with Nexia's required modules.

ryannazaretian committed:
Pinned Nexia requirements in manifest.json

ryannazaretian committed:
Reverting changes to requirements_all.txt

ryannazaretian committed:
Updated requirements_all.py from Py3 script.

homeassistant commented:
Hi @docBliny, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24947).

Once you do that we will be able to review and accept this pull request.

Thanks!

ryannazaretian committed:
Fixed line length issue.

homeassistant commented:
Hi @docBliny, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24947).

Once you do that we will be able to review and accept this pull request.

Thanks!

ryannazaretian committed:
Fixed several formatting issues found by static-check

homeassistant commented:
Hi @docBliny, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24947).

Once you do that we will be able to review and accept this pull request.

Thanks!

ryannazaretian committed:
More fixes for static-check.

ryannazaretian committed:
More static-check fixes.

Getting SyntaxError for unknown reason. Think it may be due to possible conversion to str issue.

ryannazaretian commented:
I'm not sure why the static-check is finding syntax errors. There's nothing wrong with the code.

pvizeli commented:
Look like that is not allow: https://github.com/home-assistant/architecture/blob/master/adr/0004-webscraping.md

balloob commented:
Pascal is correct. We shouldn't rely on webscraping and any protocol related code should be stored inside a third party PyPI lib. 

Feel free to promote this integration as a custom component, but we no longer rely on webscraping for any of the integrations in our repository.
